### PR TITLE
[FIX] Sqlite3 Should create db for missing parent directories as well

### LINKED ISSUE #51623: `ActiveRecord::ConnectionAdapters::SQLite3Adapter#initialize` does not correctly create missing parent directories
[`ActiveRecord::ConnectionAdapters::SQLite3Adapter#initialize`](https://github.com/rails/rails/blob/284baa18c2c62abec170b61e285aefbd9604f675/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L112-L122) will check if a sqlite3 database file path exists, and if not it will attempt to create the directory using `Dir.mkdir`. However, `Dir.mkdir` cannot create *all* parent directories like `FileUtils.mkdir_p`. Thus an `No such file or directory @ dir_s_mkdir - does/not/exist/yet (Errno::ENOENT)` will be raised if one of the parent-parent directories of the sqlite3 database file do not exist yet.

### Steps to reproduce

1. `ActiveRecord::Base.establish_connect({adapter: 'sqlite3', database: 'does/not/exist/yet/database.sqlite3'})`
2. `migrations = ActiveRecord::MigrationContext.new(['path/to/db/migrate'])`

# Your reproduction script goes here

```ruby
require 'active_record'

ActiveRecord::Base.establish_connect({adapter: 'sqlite3', database: 'does/not/exist/yet/database.sqlite3'})
migrations = ActiveRecord::MigrationContext.new(['path/to/db/migrate'])
```

### Expected behavior

Use `FileUtils.mkdir_p` to create all parent directories for the missing sqlite3 database file.

### Actual behavior

```
/home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/sqlite3_adapter.rb:119:in `rescue in initialize': Database not found (ActiveRecord::NoDatabaseError)
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/sqlite3_adapter.rb:115:in `initialize'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/sqlite3_adapter.rb:24:in `new'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/sqlite3_adapter.rb:24:in `sqlite3_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:676:in `public_send'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:676:in `new_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:723:in `checkout_new_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:702:in `try_to_checkout_new_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:654:in `acquire_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:353:in `checkout'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:182:in `connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_handler.rb:246:in `retrieve_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_handling.rb:287:in `retrieve_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_handling.rb:254:in `connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/tasks/database_tasks.rb:510:in `migration_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/migration.rb:1367:in `connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/migration.rb:1229:in `initialize'
	... 5 levels...
/home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/sqlite3_adapter.rb:116:in `mkdir': No such file or directory @ dir_s_mkdir - does/not/exist/yet (Errno::ENOENT)
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/sqlite3_adapter.rb:116:in `initialize'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/sqlite3_adapter.rb:24:in `new'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/sqlite3_adapter.rb:24:in `sqlite3_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:676:in `public_send'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:676:in `new_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:723:in `checkout_new_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:702:in `try_to_checkout_new_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:654:in `acquire_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:353:in `checkout'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:182:in `connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/connection_handler.rb:246:in `retrieve_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_handling.rb:287:in `retrieve_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/connection_handling.rb:254:in `connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/tasks/database_tasks.rb:510:in `migration_connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/migration.rb:1367:in `connection'
	from /home/postmodern/.gem/ruby/gems/activerecord-7.1.3.2/lib/active_record/migration.rb:1229:in `initialize'
	... 5 levels...
```

### System configuration

**Rails version**: `7.1.3.2`

**Ruby version**: `ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux]`

### PR DESCRIPTION
**Pull request by maniSHarma7575**
### Motivation / Background

This Pull Request has been created to fix #51623 

[ActiveRecord::ConnectionAdapters::SQLite3Adapter#initialize](https://github.com/rails/rails/blob/284baa18c2c62abec170b61e285aefbd9604f675/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L112-L122) will check if a sqlite3 database file path exists, and if not it will attempt to create the directory using Dir.mkdir. However, Dir.mkdir cannot create all parent directories like FileUtils.mkdir_p. Thus an No such file or directory @ dir_s_mkdir - does/not/exist/yet (Errno::ENOENT) will be raised if one of the parent-parent directories of the sqlite3 database file do not exist yet.

### Detail

This Pull Request changes:
ActiveRecord::ConnectionAdapters::SQLite3Adapter#initialize

Replaced the `Dir.mkdir(dirname)` with `FileUtils.mkdir_p(dirname)` to create the directory for the parent directories in the path if they don't exists.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #51623 ]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by maniSHarma7575** [comment]
@fatkodima this is ready for review? Please see if we can merge this.

File: activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb [hunk 1]
```diff
@@ -23,12 +23,12 @@ def setup
         )
       end
 
-      def test_bad_connection
```
Comment (byroot) [inline activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: This test shouldn't be deleted but updated so it still simulate a failure. Of course it's trickier now that `mkdir_p` is used.

One way could be to create a temporary directory and change its permissions. e.g.

```ruby
dir = Dir.mktmpdir
File.chmod(0x000, dir)
db_path  = File.join(dir, "db/cant-be-created.db")
```

Comment (maniSHarma7575) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: @byroot, observe the alterations made:

**Before:**

```ruby
begin
  Dir.mkdir(dirname)
rescue ` => error
  if error.message.include?("No such file or directory")
    raise ActiveRecord::NoDatabaseError.new(connection_pool: @pool)
  else
    raise
  end
end
```

Now:

```ruby
  FileUtils.mkdir_p(dirname)
```

The removal of the rescue block is notable. Why?

1. The likelihood of encountering an `Errno::ENOENT` exception diminishes since the file or directory will now be created.
2. In scenarios where permission constraints prevent directory or file creation, triggering a failure, an `Errno::EACCES` exception will be raised instead.

Incorporating a failure test, I've implemented a test here `activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:16`

```ruby
def test_database_should_not_get_created_for_parent_directories_with_missing_permission
  dir = Dir.mktmpdir
  File.chmod(0x000, dir)
  db_path = File.join(dir, "db/cant-be-created.db")
  assert_raise Errno::EACCES do
    connection = SQLite3Adapter.new(adapter: "sqlite3", database: db_path)
    connection.drop_table "ex", if_exists: true
  end
end
```

This test is green on my local machine. And it's strange that in buildkit CI it's failing. Could you please check here what I am doing wrong?

```
bin/test test/cases/adapters/sqlite3/sqlite3_adapter_test.rb
Using sqlite3
Run options: --seed 52964

# Running:

..................................................................................

Finished in 0.070632s, 1160.9469 runs/s, 2718.3146 assertions/s.
82 runs, 192 assertions, 0 failures, 0 errors, 0 skips
```

Comment (byroot) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: Not sure, try `File.chmod(0x000, dir)` (not `FileUtils`).

Also the bubbled exception shouldn't be an errno, we should handle all `SyscallError` and wrap them in `ActiveRecord::NoDatabaseError`.

Comment (maniSHarma7575) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: Thanks, @byroot! I also attempted to use `File.chmod(0x000, dir)`.
https://github.com/rails/rails/pull/51628/files#diff-91ca39a63c4ff81c0580c12472a0a709c1d9c428a3ace4970ff3076d90ba6b13

However, it isn't altering the directory's permissions.
Interestingly, the same test runs fine on my local machine.

Comment (maniSHarma7575) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: @byroot , I've identified the issue above and why it's exhibiting this unusual behavior.

To replicate the behavior, I conducted independent experiments with the FileUtils.mkdir_p method. Consequently, I crafted a script to test the method:

```ruby
require 'test/unit'
require 'fileutils'
require 'tmpdir'
require 'debug'

class TestFileUtils < Test::Unit::TestCase
  def test_write_independent_test_to_check_fileutils
    Dir.mktmpdir do |dir|
      # Create a directory with restricted permissions
      File.chmod(0o000, dir)

      # Attempt to create a directory within the restricted directory
      db_path = File.join(dir, "check-fileutils/db/-cinco-dog.sqlite3")

      assert_raise Errno::EACCES do
        FileUtils.mkdir_p(db_path)
      end
    end
  end
end
```

When I executed the above script on my local machine with Ruby installed (`ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin22]`), it ran without issues:

```shell
(base)  manishsharma@Manishs-MacBook-Air  ~/Development/personal/opensource/experiment  ruby reproduce.rb
Loaded suite reproduce
Started
.
Finished in 0.000621 seconds.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1 tests, 1 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1610.31 tests/s, 1610.31 assertions/s
```

Since Builtkite Rails CI runs on Docker, I decided to run the same script within a Docker container. For this purpose, I created the following Dockerfile:

```dockerfile
FROM ruby:latest
COPY reproduce.rb .
ENTRYPOINT ["ruby", "reproduce.rb"]
```

The `reproduce.rb` script remains unchanged.

Here's the output from running and building the image:

```shell
(base)  manishsharma@Manishs-MacBook-Air  ~/Development/personal/opensource/experiment  docker run fileutils-reproduce
Loaded suite reproduce
Started
F
===============================================================================
Failure: test_write_independent_test_to_check_fileutils(TestFileUtils): <Errno::EACCES> exception was expected but none was thrown.
reproduce.rb:15:in `block in test_write_independent_test_to_check_fileutils'
     12:       # Attempt to create a directory within the restricted directory
     13:       db_path = File.join(dir, "check-fileutils/db/-cinco-dog.sqlite3")
     14:
  => 15:       assert_raise Errno::EACCES  do
     16:         FileUtils.mkdir_p(db_path)
     17:       end
     18:     end
/usr/local/lib/ruby/3.3.0/tmpdir.rb:99:in `mktmpdir'
reproduce.rb:8:in `test_write_independent_test_to_check_fileutils'
===============================================================================
Finished in 0.003941125 seconds.
-------------------------------------------------------------------------------
1 tests, 1 assertions, 1 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
0% passed
-------------------------------------------------------------------------------
253.73 tests/s, 253.73 assertions/s
```

**The question arises: Why does the Docker environment not raise an exception?**

Upon inspecting the Docker container, I discovered that `chmod` sets the correct permissions for the directory:

```shell
# pwd
/tmp
# ls -la
total 16
drwxrwxrwt 1 root root 4096 Apr 27 15:07 .
drwxr-xr-x 1 root root 4096 Apr 27 15:03 ..
d--------- 2 root root 4096 Apr 27 15:03 d20240427-1-ihvptr
d--------- 3 root root 4096 Apr 27 15:09 d20240427-24-sa4xz6
```

Creating another directory inside `d20240427-24-sa4xz6` occurs without errors because both the owner and group have root privileges:

``` shell
# cd d20240427-24-sa4xz6
# mkdir -p create-without-issue
# ls -la
total 16
d--------- 4 root root 4096 Apr 27 15:27 .
drwxrwxrwt 1 root root 4096 Apr 27 15:07 ..
drwxr-xr-x 2 root root 4096 Apr 27 15:27 create-without-issue
drwxr-xr-x 2 root root 4096 Apr 27 15:09 temp
```

**In summary, the root user can create files in a directory marked as Read Only.**

Comment (byroot) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: I'll have a deeper look Wednesday.

Comment (manishElitmus) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: @byroot Did you had a chance to look at it?

Comment (maniSHarma7575) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: **Tried solution:**

1. Make directories immutable (using chattr or chflags) 

with chattr command it gives an error:
```
chattr: Operation not permitted while setting flags on <directory-name>
```

**Outcome:** Docker Deny access to some filesystem operations, like creating new device nodes, changing the owner of files, or altering attributes (including the immutable flag).

It's mentioned here: https://docs.docker.com/engine/security/#linux-kernel-capabilities

Solution to solve this is: This is related to capabilities thing: chattr requires CAPLINUX IMMUTABLE which is disabled in docker by default. Just add -- cap-add
LINUX_ IMMUTABLE to docker container start options to enable it.

That's itself another issue on. I don't know whether to proceed with this or not as we have to do modification in CI.

Comment (byroot) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: Sorry, got side tracked, looking at it now.

Comment (byroot) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: Alright, I think that the issue is CI runs as root, so it's hard to simulate any problem into creating a directory or file.

So while I don't like not covering this sort of things with a test, I don't quite see what we could do, and anyway it could likely fail in dozens of different ways, so coverage here will never be that good.

I'll remove the test, sorry for sending you on a goose chase.

Comment (maniSHarma7575) [reply activerecord/test/cases/adapters/sqlite3/sqlite3_adapter_test.rb:26]: Thanks! @byroot. Goose chase helped me go dig deep into docker and linux.
